subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -228,78 +300,143 @@ static void firmware_transfer(struct k_work *work)
+	len = parsed_uri.field_data[UF_SCHEMA].len;
+	if (len != 4 || memcmp(firmware_uri + off, "coap", 4)) {
+		SYS_LOG_ERR("Unsupported schema");
+		lwm2m_firmware_set_update_result(RESULT_UNSUP_PROTO);
+		return;
+	}
+
+	if (!(parsed_uri.field_set & (1 << UF_PORT))) {
+		/* Set to default port of CoAP */
+		parsed_uri.port = 5683;
+	}
+
+	off = parsed_uri.field_data[UF_HOST].off;
+	len = parsed_uri.field_data[UF_HOST].len;
+
+	/* IPv6 is wrapped by brackets */
+	if (off > 0 && firmware_uri[off - 1] == '[') {
+		if (!IS_ENABLED(CONFIG_NET_IPV6)) {
+			SYS_LOG_ERR("Doesn't support IPv6");
+			lwm2m_firmware_set_update_result(RESULT_UNSUP_PROTO);
+			return;
+		}
+
+		family = AF_INET6;
+	} else {
+		/* Distinguish IPv4 or DNS */
+		for (int i = off; i < off + len; i++) {
+			if (!isdigit(firmware_uri[i]) &&
+			    firmware_uri[i] != '.') {
+				SYS_LOG_ERR("Doesn't support DNS lookup");
+				lwm2m_firmware_set_update_result(
+						RESULT_UNSUP_PROTO);
+				return;
+			}
+		}
+
+		if (!IS_ENABLED(CONFIG_NET_IPV4)) {
+			SYS_LOG_ERR("Doesn't support IPv4");
+			lwm2m_firmware_set_update_result(RESULT_UNSUP_PROTO);
+			return;
+		}
+
+		family = AF_INET;
+	}
+
+	tmp = firmware_uri[off + len];
+	firmware_uri[off + len] = '\0';
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
 		firmware_addr.sa_family = family;
-		/* HACK: use firmware_uri directly as IP address */
-		net_addr_pton(firmware_addr.sa_family, firmware_uri,
+		net_addr_pton(firmware_addr.sa_family, firmware_uri + off,
 			      &net_sin6(&firmware_addr)->sin6_addr);
-		net_sin6(&firmware_addr)->sin6_port = htons(5685);
+		net_sin6(&firmware_addr)->sin6_port = htons(parsed_uri.port);
 	}
-#endif
 
-#if defined(CONFIG_NET_IPV4)
-	if (family == AF_INET) {
+	if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
 		firmware_addr.sa_family = family;
-		net_addr_pton(firmware_addr.sa_family, firmware_uri,
+		net_addr_pton(firmware_addr.sa_family, firmware_uri + off,
 			      &net_sin(&firmware_addr)->sin_addr);
-		net_sin(&firmware_addr)->sin_port = htons(5685);
+		net_sin(&firmware_addr)->sin_port = htons(parsed_uri.port);
 	}
-#endif
+
+	/* restore firmware_uri */
+	firmware_uri[off + len] = tmp;
 
 	ret = net_context_get(firmware_addr.sa_family, SOCK_DGRAM, IPPROTO_UDP,
 			      &firmware_net_ctx);
 	if (ret) {
-		NET_ERR("Could not get an UDP context (err:%d)", ret);
+		SYS_LOG_ERR("Could not get an UDP context (err:%d)", ret);
+		lwm2m_firmware_set_update_result(RESULT_CONNECTION_LOST);
 		return;
 	}
 
 	iface = net_if_get_default();
 	if (!iface) {
-		NET_ERR("Could not find default interface");
-		goto cleanup;
+		SYS_LOG_ERR("Could not find default interface");
+		lwm2m_firmware_set_update_result(RESULT_CONNECTION_LOST);
+		return;
 	}
 
-#if defined(CONFIG_NET_IPV6)
-	if (firmware_addr.sa_family == AF_INET6) {
+	if (IS_ENABLED(CONFIG_NET_IPV6) && firmware_addr.sa_family == AF_INET6) {
 		ret = net_context_bind(firmware_net_ctx,
 				       (struct sockaddr *)&any_addr6,
```

Comment:
Isn't there now a compile error if you disable IPv6 as the any_addr6 is only found if IPv6 is enabled?